### PR TITLE
Adding copy flow function plus some cosmetic changes

### PR DESCRIPTION
**Pull request by ifilippov**

### CODE REVIEW

File: flow/flow.go [hunk 1]
```diff
@@ -557,8 +574,8 @@ func SetGenerator(generateFunction interface{}, targetSpeed uint64, context User
 func SetSender(IN *Flow, par interface{}) {
 	checkFlow(IN)
 	var send *scheduler.FlowFunction
-	if port, t := par.(uint8); t {
-		if port >= uint8(len(createdPorts)) {
+	if port, t := par.(int); t {
+		if uint8(port) >= uint8(len(createdPorts)) {
```
Comment (dkolistr) [inline flow/flow.go]: do we really need conversion of ints to uints8 here?

File: flow/flow.go [hunk 2]
```diff
@@ -496,8 +513,8 @@ func SetReader(filename string, repcount int32) (OUT *Flow) {
 func SetReceiver(par interface{}) (OUT *Flow) {
 	ring := low.CreateRing(generateRingName(), burstSize*sizeMultiplier)
 	var recv *scheduler.FlowFunction
-	if port, t := par.(uint8); t {
-		if port >= uint8(len(createdPorts)) {
+	if port, t := par.(int); t {
```
Comment (mbulatova) [inline flow/flow.go]: If int here, also need to modify files examples/Firewall.go and other - remove port conversion to uint8

File: test/stash/create_packet_example.go [hunk 3]
```diff
@@ -41,7 +41,7 @@ func main() {
 }
 
 func generatePacket(pkt *packet.Packet, context flow.UserContext) {
-	// Total packet size will be 14+20+20+70+4(crc)=128 bytes
+	// Total packet size will be 14+70+4(crc)=128 bytes
```
Comment (dkolistr) [inline test/stash/create_packet_example.go]: 14+70+4 != 128...

File: low/low.go [hunk 4]
```diff
@@ -503,8 +503,8 @@ func SetAffinity(coreID uint8) {
 }
 
 // AllocateMbufs allocates a bulk of mbufs.
```
Comment (dkolistr) [inline low/low.go]: maybe comment about n? why is this change needed?

File: flow/flow.go [hunk 5]
```diff
@@ -156,6 +156,23 @@ func makeSender(port uint8, queue int16, in *low.Ring) *scheduler.FlowFunction {
 	return schedState.NewUnclonableFlowFunction("sender", ffCount, send, par)
 }
 
+type copyParameters struct {
+	in      *low.Ring
+	out1    *low.Ring
```
Comment (dkolistr) [inline flow/flow.go]: I'm caviling at it...but maybe out1 and out2 would better be out and outCopy?

**Comment by dkolistr** [comment]
You changed function signature in low.go, but did not change its usage in some files (like major tests in packet package for instance) it is a baaad-baaad practice :)

**Comment by ifilippov** [comment]
Good catch, thank you!

**Comment by ifilippov** [comment]
Updated

File: test/stability/reordering/recv.go [hunk 6]
```diff
@@ -19,7 +19,7 @@ const (
 
 type stats struct {
 	count, reordered uint64
-	duration time.Duration
+	duration         time.Duration
```
[the remark this answers is not in the document]
Comment (ifilippov) [reply test/stability/reordering/recv.go]: This is gofmt, I suppose that because of previous line with two variables

File: low/low.go [hunk 7]
```diff
@@ -502,9 +502,9 @@ func SetAffinity(coreID uint8) {
 	syscall.RawSyscall(syscall.SYS_SCHED_SETAFFINITY, uintptr(0), unsafe.Sizeof(cpuset), uintptr(unsafe.Pointer(&cpuset)))
 }
 
-// AllocateMbufs allocates a bulk of mbufs.
-func AllocateMbufs(mb []uintptr, mempool *Mempool) {
-	err := C.allocateMbufs((*C.struct_rte_mempool)(mempool), (**C.struct_rte_mbuf)(unsafe.Pointer(&mb[0])), C.unsigned(len(mb)))
+// AllocateMbufs allocates n mbufs.
+func AllocateMbufs(mb []uintptr, mempool *Mempool, n uint) {
```
[the remark this answers is not in the document]
Comment (ifilippov) [reply low/low.go]: We want this. This change is just to be able to allocate X and fill less than X

File: packet/arp_test.go [hunk 8]
```diff
@@ -33,7 +33,7 @@ var (
 func TestInitARPCommonDataacket(t *testing.T) {
 	// Create empty packet, set UDP header fields
 	mb := make([]uintptr, 1)
-	low.AllocateMbufs(mb, mempoolARPTest)
+	low.AllocateMbufs(mb, mempoolARPTest, 1)
```
[the remark this answers is not in the document]
Comment (ifilippov) [reply packet/arp_test.go]: It is tests constants from above line
